Python/Product/Analysis/EncodedLocation.cs

```diff
@@ -39,16 +39,16 @@ struct EncodedLocation : IEquatable<EncodedLocation> {
         public readonly object Location;
 
         public EncodedLocation(ILocationResolver resolver, object location) {
+            if (resolver == null && !(location is LocationInfo)) {
+                throw new ArgumentNullException(nameof(resolver));
+            }
+
             Resolver = resolver;
             Location = location;
         }
 
         public override int GetHashCode() {
-            if (Location != null) {
-                return Resolver.GetHashCode() ^ Location.GetHashCode();
-            }
-
-            return Resolver.GetHashCode();
+            return (Resolver?.GetHashCode() ?? 0) ^ (Location?.GetHashCode() ?? 0);
```

Comment:
Yep. `GetHashCode` has to be valid for any value, so tuple has to do whatever magic it does to take care of producing something when some of the elements are null.

The main benefit of using tuples for this, aside from conciseness, is that you get a much better implementation that combines constituent hashes - there are many known scenarios where `^` produces pathologically bad hashes.